Sources/Datadog/URLSessionAutoInstrumentation/URLSessionSwizzler.swift

```diff
@@ -0,0 +1,205 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+internal class URLSessionSwizzler {
+    /// `URLSession.dataTask(with:completionHandler:)` (for `URLRequest`) swizzling.
+    let dataTaskWithURLRequestAndCompletion: DataTaskWithURLRequestAndCompletion
+    /// `URLSession.dataTask(with:)` (for `URLRequest`) swizzling.
+    let dataTaskWithURLRequest: DataTaskWithURLRequest
+
+    /// `URLSession.dataTask(with:completionHandler:)` (for `URL`) swizzling. Only applied on iOS 13 and above.
+    let dataTaskWithURLAndCompletion: DataTaskWithURLAndCompletion?
+    /// `URLSession.dataTask(with:)` (for `URL`) swizzling. Only applied on iOS 13 and above.
+    let dataTaskWithURL: DataTaskWithURL?
+
+    private let interceptor: URLSessionInterceptorType
+
+    init(interceptor: URLSessionInterceptorType) throws {
+        if #available(iOS 13.0, *) {
+            self.dataTaskWithURLAndCompletion = try DataTaskWithURLAndCompletion(interceptor: interceptor)
+            self.dataTaskWithURL = try DataTaskWithURL(interceptor: interceptor)
+        } else {
+            // Prior to iOS 13.0 we do not apply following swizzlings, as those methods call
+            // the `URLSession.dataTask(with:completionHandler:)` internally which is managed
+            // by the `DataTaskWithURLRequestAndCompletion` swizzling.
```

Comment:
This internal behaviour of `session.dataTask(with:)` is only different prior to iOS 13.0 (iOS 11.x-12.x) - on iOS 14.0 it works the same as on iOS 13.x. 

Yes, we don't know how this will behave in future versions of the iOS, but we cannot also assume that "one approach to fit iOS 11.0 - 14.0" will work for iOS 15 as well.

As we depend on the `URLSession` internal details here, I prefer the approach of explicitly handling known cases, instead of trying to write single solution that will cover them all and hope that it will remain correct in the future. With this approach, we can hide all known `URLSession` details in `swizzler` implementation, and not leak it to the `interceptor` as discussed in https://github.com/DataDog/dd-sdk-ios/pull/262#discussion_r500103775.

Anyway, this is convention, I'm wondering WDYT @mariusc83 @xgouchet ?

Reply:
yes sure we should handle both known cases but i mean in this particular case i would try to avoid `#if available(...)` conditions.
this too will work certainly at least until iOS 15 ✅ 
i'm resolving the convo but @xgouchet @mariusc83 feel free to add comments

Reply:
The way we solve such situations in Android is that we deal with the current use cases (API versions) and we improve later if needed with up - coming Android versions. 
For example a real scenario would be the way we are dealing with the View/Gesture auto - instrumentation. All the strategies that we provided so far are covering the current common cases (Activity, Fragments). We are not covering yet the `Composable` API which will be a new way of implementing your Views. This API is still in Alpha and until we do not see a real adoption tendance we are not going to spend the time to provide any auto - instrumentation strategy for it. Bottom line is keep it simple, cover what you can for what you have because you will definitely change that code there at some point and you definitely cannot guess a magic formula which will cover all the edge cases now.

Reply:
The way I see it, we target the latest available OS versions, and only handle case when unsupported/required by older version. Also whenever a new version of an OS is made available to developers, we'll schedule a task early on to ensure we support it and support each corner case as they arise.
So for the current topic, we target iOS 14/13 so let's keep that as the default, and as was done here do something else for the lower version. In general we can assume the behavior will stay the same in future version unless we see it changing when those future version arrive.